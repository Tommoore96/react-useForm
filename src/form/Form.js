import React from 'react'

const Form = ({ children, ...rest }) => {
  return (
    <form
      {...rest}
      onSubmit={e => {
        e.preventDefault()
      }}
    >
      {children}
    </form>
  )
}

export default Form
