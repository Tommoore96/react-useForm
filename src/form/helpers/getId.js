export const getId = (parent, key)=> {
  if(parent) return `${parent}-${key}`
  return key
}
