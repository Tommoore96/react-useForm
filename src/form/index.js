export { useForm } from './useForm'
export { createMemoryForm } from '../memoryForm/createMemoryForm'
