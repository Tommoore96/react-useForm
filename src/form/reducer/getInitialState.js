import { fromJS, Map, List } from 'immutable'
import { fieldsKey, listenersKey } from './fieldReducer'
import { getFieldState } from './getFieldState'

/**
 * @param fields {Array<import("..").IFormFieldArgs>}
 */
export const getInitialState = (fields, initialValues = Map(), options = {}, formTools) => {
  const fieldMap = fields.reduce((acc, field) => {
    const result = acc.set(field.name, getFieldState(field, initialValues, options))

    return result
  }, Map())

  const listeners = getListeners(options.listeners)
  const withFormChange = options.onFormChange ? fromJS({ onFormChange: options.onFormChange }) : Map()
  return fromJS({ [fieldsKey]: fieldMap })
    .merge(listeners)
    .merge(withFormChange)
    .set('formTools', formTools)
}

const getListeners = (listeners = {}) => {
  return Object.entries(listeners).reduce((acc, [fieldName, value]) => {
    return acc.updateIn([listenersKey, fieldName], List(), current => {
      return Array.isArray(value) ? current.concat(value) : current.push(value)
    })
  }, Map())
}
