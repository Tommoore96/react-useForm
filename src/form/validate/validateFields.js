import { Map } from 'immutable'
import { prepareNameForValidate } from '../fields/prepareNameForValidate'

const getListValues = items => {
  return items.map(item => {
    return item.get('fields').entrySeq().reduce((acc, [key, value]) => {
      return acc.set(key, value.getIn(['current', 'value']))
    }, Map())
  })
}

const validateItems = (items, getAllValues) => {
  return items.map(item => {
    return validateMap(item.get('fields'), getAllValues)
  }).toArray()
}

const validateMap = (fields, getAllValues) => {
  return fields.entrySeq().reduce((acc, [key, field]) => {
    const value = field.getIn(['current', 'value'])
    const optional = field.getIn(['initial', 'optional'], false)
    const { validate, requiredMessage, name } = field.getIn(['initial', 'field'], {})

    const items = field.get('items')
    if (items) {
      acc[key] = validateItems(items, getAllValues)
      if (!optional && items.size === 0) {
        acc[key].error = requiredMessage || 'Required'
      }

      if (validate) {
        const error = validate(getListValues(items), prepareNameForValidate(name), getAllValues)
        if (error) {
          acc[key].error = error
        }
      }


      return acc
    }

    if (validate) {
      const error = validate(value, prepareNameForValidate(name), getAllValues)
      if (error) {
        acc[key] = error
        return acc
      }
    }

    if (!optional && (value === undefined || value === '')) {
      acc[key] = requiredMessage || 'Required'
    }

    return acc
  }, {})
}

export const validateFields = (state, getAllValues) => {
  return validateMap(state.get('fields'), getAllValues)
}
