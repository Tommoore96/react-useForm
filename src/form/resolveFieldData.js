import { Map } from 'immutable'
import { fieldsKey } from './reducer/fieldReducer'
import {
  useBooleanField,
  useNumberField,
  useListField,
  useSelectField,
  useTextField,
  useObjectField,
} from './fields'
import {getId} from './helpers/getId'

export const resolveFieldData = (state, dispatch, getAllValues, formOptions) => {
  const fieldData = state.get(fieldsKey, Map()).entrySeq().reduce((acc, [fieldName, current]) => {
    const fieldState = current.set('getAllValues', getAllValues)

    acc[fieldName] = resolveField(fieldState, dispatch, getAllValues, formOptions)
    return acc
  }, {})

  return fieldData
}

export const resolveField = (fieldState, dispatch, getAllValues, formOptions) => {
  const field = fieldState.getIn(['initial', 'field'])
  const fieldType = fieldState.getIn(['initial', 'type'])
  switch (fieldType) {
    case 'select':
      return useSelectField(fieldState, dispatch, field)
    case 'boolean':
      return useBooleanField(fieldState, dispatch, field)
    case 'number':
      return useNumberField(fieldState, dispatch, field)
    case 'list':
      return useListField(fieldState, dispatch, field, getAllValues, formOptions)
    case 'object':
      return useObjectField(fieldState, dispatch, field)
    case 'text':
    default:
      return useTextField(fieldState, dispatch, field)
  }
}

export const getFieldProps = (fieldData, state = Map(), parentKey = '') => {
  return Object.entries(fieldData).reduce((acc, [key, v]) => {
    const passThrough = state.getIn([fieldsKey, key, 'initial', 'field'], {}).passThrough || {}

    acc[key] = {
      id: getId(parentKey, key),
      ...v.props,
      ...passThrough,
    }
    return acc
  }, {})
}
