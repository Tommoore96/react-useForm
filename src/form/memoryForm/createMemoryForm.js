import { fromJS } from 'immutable'
import { getInitialState } from '../reducer/getInitialState'
import { validateAll } from '../validate/validateAll'
import { resolveFieldData } from '../resolveFieldData'
import { getFieldValues } from '../helpers/getFieldValues'
import { pruneNonErrors } from '../validate/pruneNonErrors'

export const createMemoryForm = ({ fields, validate, initialValues }) => {
  let fieldData
  const formTools = {
    current: {},
  }
  const fakeDispatch = () => undefined
  const getAllValues = () => fromJS(getFieldValues(fieldData))

  const state = getInitialState(fields, initialValues, {}, formTools)
  fieldData = resolveFieldData(state, fakeDispatch, getAllValues)

  const validateForm = () => {
    let errors = validateAll(state, fieldData, validate, getAllValues)
    errors = pruneNonErrors(errors)
    return {
      errors,
      isValid: Object.keys(errors).length === 0,
    }
  }

  return {
    validate: validateForm,
  }
}
