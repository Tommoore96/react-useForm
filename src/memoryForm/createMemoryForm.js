import { fromJS } from 'immutable'
import { getInitialState } from '../form/reducer/getInitialState'
import { validateAll } from '../form/validate/validateAll'
import { resolveFieldData } from '../form/resolveFieldData'
import { getFieldValues } from '../form/helpers/getFieldValues'
import { pruneNonErrors } from '../form/validate/pruneNonErrors'

export const createMemoryForm = ({ fields, validate, initialValues }) => {
  let fieldData
  const formTools = {
    current: {},
  }
  const fakeDispatch = () => undefined
  const getAllValues = () => fromJS(getFieldValues(fieldData))

  const state = getInitialState(fields, initialValues, {}, formTools)
  fieldData = resolveFieldData(state, fakeDispatch, getAllValues)

  const validateForm = () => {
    let errors = validateAll(state, fieldData, validate, getAllValues)
    errors = pruneNonErrors(errors)
    return {
      errors,
      isValid: Object.keys(errors).length === 0,
    }
  }

  return {
    validate: validateForm,
  }
}
