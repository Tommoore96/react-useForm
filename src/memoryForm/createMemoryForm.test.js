import { fromJS, Map } from 'immutable'
import { getInitialState } from '../form/reducer/getInitialState'
import { validateAll } from '../form/validate/validateAll'
import { resolveFieldData } from '../form/resolveFieldData'
import { createMemoryForm } from './createMemoryForm'
import { getFieldValues } from '../form/helpers/getFieldValues'
import { pruneNonErrors } from '../form/validate/pruneNonErrors'

jest.mock('../form/reducer/getInitialState')
jest.mock('../form/validate/validateAll')
jest.mock('../form/resolveFieldData')
jest.mock('../form/helpers/getFieldValues')
jest.mock('../form/validate/pruneNonErrors')

let fields, validators, initialValues
beforeEach(() => {
  fields = [
    { name: 'name', type: 'text', label: 'Name' },
    { name: 'address', type: 'text', label: 'Address' },
  ]

  validators = ['validator1']

  initialValues = fromJS({ 'name': 'the-name', address: 'the-address' })

  getInitialState.mockReturnValue(initialValues.merge(Map({ 'resolvedInitialState': true })))
  resolveFieldData.mockReturnValue('field-data')
  validateAll.mockReturnValue({ isValid: true, errors: {} })
  getFieldValues.mockReturnValue({ name: 'name', address: 'address' })
  pruneNonErrors.mockReturnValue({ errors: 'pruned' })
})

test('gets initial state from fields and values', () => {
  createMemoryForm({ fields, validate: validators, initialValues })

  expect(getInitialState.mock.calls[0]).toMatchSnapshot()
})

test('gets field data', () => {
  createMemoryForm({ fields, validate: validators, initialValues })

  expect(resolveFieldData.mock.calls[0]).toMatchSnapshot()
})

test('returns form functions', () => {
  expect(createMemoryForm({ fields, validate: validators, initialValues })).toMatchSnapshot()
})

test('validates the form', () => {
  const { validate } = createMemoryForm({ fields, validate: validators, initialValues })

  expect(validate()).toMatchSnapshot()
  expect(validateAll.mock.calls[0]).toMatchSnapshot()
})

test('validation: getAllValues returns getFieldValues', () => {
  const { validate } = createMemoryForm({ fields, validate: validators, initialValues })

  validate()

  const actual = validateAll.mock.calls[0][3]()

  expect(getFieldValues.mock.calls[0]).toMatchSnapshot()
  expect(actual).toMatchSnapshot()
})
