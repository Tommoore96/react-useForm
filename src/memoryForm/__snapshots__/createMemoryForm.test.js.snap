// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets field data 1`] = `
Array [
  Immutable.Map {
    "name": "the-name",
    "address": "the-address",
    "resolvedInitialState": true,
  },
  [Function],
  [Function],
]
`;

exports[`gets initial state from fields and values 1`] = `
Array [
  Array [
    Object {
      "label": "Name",
      "name": "name",
      "type": "text",
    },
    Object {
      "label": "Address",
      "name": "address",
      "type": "text",
    },
  ],
  Immutable.Map {
    "name": "the-name",
    "address": "the-address",
  },
  Object {},
  Object {
    "current": Object {},
  },
]
`;

exports[`returns form functions 1`] = `
Object {
  "validate": [Function],
}
`;

exports[`validates the form 1`] = `
Object {
  "errors": Object {
    "errors": "pruned",
  },
  "isValid": false,
}
`;

exports[`validates the form 2`] = `
Array [
  Immutable.Map {
    "name": "the-name",
    "address": "the-address",
    "resolvedInitialState": true,
  },
  "field-data",
  Array [
    "validator1",
  ],
  [Function],
]
`;

exports[`validation: getAllValues returns getFieldValues 1`] = `
Array [
  "field-data",
]
`;

exports[`validation: getAllValues returns getFieldValues 2`] = `
Immutable.Map {
  "name": "name",
  "address": "address",
}
`;
